File: proj-android/PowerAuthLibrary/src/main/java/io/getlime/security/powerauth/sdk/PowerAuthAuthentication.java
/*
 * Copyright 2017 Wultra s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.getlime.security.powerauth.sdk;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import io.getlime.security.powerauth.core.Password;
import io.getlime.security.powerauth.core.SecureData;
import io.getlime.security.powerauth.system.PowerAuthLog;

import java.util.Arrays;

/**
 * Class representing a multi-factor authentication object.
 */
public class PowerAuthAuthentication {
    /**
     * If set, then the biometry factor will be used.
     */
    private final @Nullable SecureData useBiometry;
    /**
     * If set, then the biometric factor will be used.
     */
    private final @Nullable PowerAuthBiometricPrompt biometricPrompt;
    /**
     * If set, then the password will be used.
     */
    private final @Nullable Password password;
    /**
     * Optional custom possession key.
     */
    private final @Nullable SecureData overriddenPossessionKey;

    /**
     * Contains {@code true} if authentication object should be used to persist activation, {@code false}
     * if object is for authentication code calculation or {@code null} if this is a legacy object with no usage
     * specified.
     */
    private final Boolean persistActivation;

    /**
     * Construct object with desired combination of factors. Such authentication object can be used
     * either to persist activation and for the authentication code calculation.
     * <p>
     * Note that you should prefer static construction functions instead of this constructor, unless
     * you have a special reason for it.
     * <p>
     * Note that the constructor internally makes a copy of provided byte arrays. This is because authentication object
     * now supports {@link #destroy()} method that clears content of such arrays. We don't want to clear instance of
     * array that is still in use somewhere else.
     *
     *
     * @param persistActivation If true, then authentication can be used to persist activation.
     * @param password If set, then knowledge factor will be used to persist activation or authentication code calculation.
     * @param biometryFactorRelatedKey If set, then biometry factor will be used to persist activation or the authentication code calculation.
     * @param overriddenPossessionKey Custom possession factor related key.
     */
    PowerAuthAuthentication(
            Boolean persistActivation,
            @Nullable Password password,
            @Nullable PowerAuthBiometricPrompt biometricPrompt,
            @Nullable SecureData biometryFactorRelatedKey,
            @Nullable SecureData overriddenPossessionKey) {
        this.useBiometry = biometryFactorRelatedKey;
        this.biometricPrompt = biometricPrompt;
        this.password = password;
        this.overriddenPossessionKey = overriddenPossessionKey;
        this.persistActivation = persistActivation;
    }

    /**
     * Make copy of provided byte array if non-null array is provided.
     * @param original Array to safely copy.
     * @return new array with copied content or null if input is null.
     */
    private static byte[] safeArrayCopy(byte[] original) {
        if (original != null) {
            return Arrays.copyOf(original, original.length);
        }
        return null;
    }

    /**
     * Make sure that the sensitive data is always wiped out from the memory.
     */
    protected void finalize() {
        releaseSensitiveData(true);
    }

    /**
     * Release all sensitive data.
     * @param fromGC If true, then this comes from Garbage Collector, otherwise from application.
     */
    private void releaseSensitiveData(boolean fromGC) {
        if (password != null && !fromGC) {          // Do not destroy Password if request comes from GC
            password.destroy();
        }
        if (useBiometry != null) {
            useBiometry.destroy();
        }
        if (overriddenPossessionKey != null) {
            overriddenPossessionKey.destroy();
        }
    }


    // Persist activation

    /**
     * Construct authentication object to persist activation with password.
     * @param password Password to set for new activation.
     * @return Authentication object constructed to persist activation with the password.
     */
    public static PowerAuthAuthentication persistWithPassword(@NonNull String password) {
        return new PowerAuthAuthentication(true, new Password(password), null, null, null);
    }

    /**
     * Construct authentication object to persist activation with password and custom key for the possession factor.
     * @param password Password to set for new activation.
     * @param overriddenPossessionKey Custom possession key to set for new activation.
     * @return Authentication object constructed to persist activation with password, with using custom key for the possession factor.
     */
    public static PowerAuthAuthentication persistWithPassword(@NonNull String password, @NonNull SecureData overriddenPossessionKey) {
        return new PowerAuthAuthentication(true, new Password(password), null, null, overriddenPossessionKey);
    }

    /**
     * Construct authentication object to persist activation with password and with biometry.
     * @param password Password to set for new activation.
     * @param biometricPrompt Prompt displayed during the biometric authentication.
     * @return Authentication object constructed to persist activation with password and biometry.
     */
    public static PowerAuthAuthentication persistWithPasswordAndBiometry(@NonNull String password, @NonNull PowerAuthBiometricPrompt biometricPrompt) {
        return new PowerAuthAuthentication(true, new Password(password), biometricPrompt, null, null);
    }

    /**
     * Construct authentication object to persist activation with password and with biometry.
     * @param password Password to set for new activation.
     * @param biometryFactorRelatedKey Biometry factor related key to set for new activation.
     * @return Authentication object constructed to persist activation with password and biometry.
     */
    public static PowerAuthAuthentication persistWithPasswordAndBiometry(@NonNull String password, @NonNull SecureData biometryFactorRelatedKey) {
        return new PowerAuthAuthentication(true, new Password(password), null, biometryFactorRelatedKey, null);
    }

    /**
     * Construct authentication object to persist activation with password, biometry and with custom key for the possession factor.
     * @param password Password to set for new activation.
     * @param biometryFactorRelatedKey Biometry factor related key to set for new activation.
     * @param overriddenPossessionKey Custom possession key to set for new activation.
     * @return Authentication object constructed to persist activation with password, with using custom key for the possession factor.
     */
    public static PowerAuthAuthentication persistWithPasswordAndBiometry(@NonNull String password, @NonNull SecureData biometryFactorRelatedKey, @NonNull SecureData overriddenPossessionKey) {
        return new PowerAuthAuthentication(true, new Password(password), null, biometryFactorRelatedKey, overriddenPossessionKey);
    }

    // core/Password variants

    /**
     * Construct authentication object to persist activation with password.
     * @param password Password to set for new activation.
     * @return Authentication object constructed to persist activation with the password.
     */
    public static PowerAuthAuthentication persistWithPassword(@NonNull Password password) {
        return new PowerAuthAuthentication(true, password, null, null, null);
    }

    /**
     * Construct authentication object to persist activation with password and custom key for the possession factor.
     * @param password Password to set for new activation.
     * @param overriddenPossessionKey Custom possession key to set for new activation.
     * @return Authentication object constructed to persist activation and password, with using custom key for the possession factor.
     */
    public static PowerAuthAuthentication persistWithPassword(@NonNull Password password, @NonNull SecureData overriddenPossessionKey) {
        return new PowerAuthAuthentication(true, password, null, null, overriddenPossessionKey);
    }

    /**
     * Construct authentication object to persist activation with password and with biometry.
     * @param password Password to set for new activation.
     * @param biometricPrompt Prompt displayed during the biometric authentication.
     * @return Authentication object constructed to persist activation with password and biometry.
     */
    public static PowerAuthAuthentication persistWithPasswordAndBiometry(@NonNull Password password, @NonNull PowerAuthBiometricPrompt biometricPrompt) {
        return new PowerAuthAuthentication(true, password, biometricPrompt, null, null);
    }

    /**
     * Construct authentication object to persist activation with password and with biometry.
     * @param password Password to set for new activation.
     * @param biometryFactorRelatedKey Biometry factor related key to set for new activation.
     * @return Authentication object constructed to persist activation with password and biometry.
     */
    public static PowerAuthAuthentication persistWithPasswordAndBiometry(@NonNull Password password, @NonNull SecureData biometryFactorRelatedKey) {
        return new PowerAuthAuthentication(true, password, null, biometryFactorRelatedKey, null);
    }

    /**
     * Construct authentication object to persist activation with password, biometry and with custom key for the possession factor.
     * @param password Password to set for new activation.
     * @param biometryFactorRelatedKey Biometry factor related key to set for new activation.
     * @param overriddenPossessionKey Custom possession key to set for new activation.
     * @return Authentication object constructed to persist activation with password, with using custom key for the possession factor.
     */
    public static PowerAuthAuthentication persistWithPasswordAndBiometry(@NonNull Password password, @NonNull SecureData biometryFactorRelatedKey, @NonNull SecureData overriddenPossessionKey) {
        return new PowerAuthAuthentication(true, password, null, biometryFactorRelatedKey, overriddenPossessionKey);
    }

    // Authenticate

    /**
     * Construct authentication object for authentication code calculation purposes. The authentication code is calculated with possession factor only.
     * @return Authentication object constructed to calculate authentication code with possession factor only.
     */
    public static PowerAuthAuthentication possession() {
        return new PowerAuthAuthentication(false, null, null, null, null);
    }

    /**
     * Construct authentication object for authentication code calculation purposes. The authentication code is calculated with possession factor only, with using custom possession key.
     * @param overriddenPossessionKey Custom possession key to use for the authentication code calculation.
     * @return Authentication object constructed to calculate authentication code with possession factor with custom possession key.
     */
    public static PowerAuthAuthentication possession(@NonNull SecureData overriddenPossessionKey) {
        return new PowerAuthAuthentication(false, null, null, null, overriddenPossessionKey);
    }

    /**
     * Construct authentication object for authentication code calculation purposes. The authentication code is calculated with possession and knowledge factors.
     * @param password Password to use for the authentication code calculation.
     * @return Authentication object constructed to calculate authentication code with possession and knowledge factors.
     */
    public static PowerAuthAuthentication possessionWithPassword(@NonNull String password) {
        return new PowerAuthAuthentication(false, new Password(password), null, null, null);
    }

    /**
     * Construct authentication object for authentication code calculation purposes. The authentication code is calculated with possession and knowledge factors, with using custom possession key.
     * @param password Password to use for the authentication code calculation.
     * @param overriddenPossessionKey Custom possession key to use for the authentication code calculation.
     * @return Authentication object constructed to calculate authentication code with possession and knowledge factors, with using custom possession key.
     */
    public static PowerAuthAuthentication possessionWithPassword(@NonNull String password, @NonNull SecureData overriddenPossessionKey) {
        return new PowerAuthAuthentication(false, new Password(password), null, null, overriddenPossessionKey);
    }

    /**
     * Construct authentication object for authentication code calculation purposes. The authentication code is calculated with possession and biometry factors.
     * @param biometricPrompt Prompt displayed during the biometric authentication.
     * @return Authentication object constructed to calculate authentication code with possession and biometry factors
     */
    public static PowerAuthAuthentication possessionWithBiometry(@NonNull PowerAuthBiometricPrompt biometricPrompt) {
        return new PowerAuthAuthentication(false, null, biometricPrompt, null, null);
    }

    /**
     * Construct authentication object for authentication code calculation purposes. The authentication code is calculated with possession and biometry factors.
     * @param biometryFactorRelatedKey Biometry key data to use for the authentication code calculation.
     * @return Authentication object constructed to calculate authentication code with possession and biometry factors
     */
    public static PowerAuthAuthentication possessionWithBiometry(@NonNull SecureData biometryFactorRelatedKey) {
        return new PowerAuthAuthentication(false, null, null, biometryFactorRelatedKey, null);
    }

    /**
     * Construct authentication object for authentication code calculation purposes. The authentication code is calculated with possession and biometry factors, with using custom possession key.
     * @param biometryFactorRelatedKey Biometry key data to use for the authentication code calculation.
     * @param overriddenPossessionKey Custom possession key to use for the authentication code calculation.
     * @return Authentication object constructed to calculate authentication code with possession and biometry factors, with using custom possession key.
     */
    public static PowerAuthAuthentication possessionWithBiometry(@NonNull SecureData biometryFactorRelatedKey, @NonNull SecureData overriddenPossessionKey) {
        return new PowerAuthAuthentication(false, null, null, biometryFactorRelatedKey, overriddenPossessionKey);
    }

    // core/Password variants
    
    /**
     * Construct authentication object for authentication code calculation purposes. The authentication code is calculated with possession and knowledge factors.
     * @param password Password to use for the authentication code calculation.
     * @return Authentication object constructed to calculate authentication code with possession and knowledge factors.
     */
    public static PowerAuthAuthentication possessionWithPassword(@NonNull Password password) {
        return new PowerAuthAuthentication(false, password, null, null, null);
    }

    /**
     * Construct authentication object for authentication code calculation purposes. The authentication code is calculated with possession and knowledge factors, with using custom possession key.
     * @param password Password to use for the authentication code calculation.
     * @param overriddenPossessionKey Custom possession key to use for the authentication code calculation.
     * @return Authentication object constructed to calculate authentication code with possession and knowledge factors, with using custom possession key.
     */
    public static PowerAuthAuthentication possessionWithPassword(@NonNull Password password, @NonNull SecureData overriddenPossessionKey) {
        return new PowerAuthAuthentication(false, password, null, null, overriddenPossessionKey);
    }

    /**
     * Determines whether the authentication code should be calculated using the biometric factor.
     * Biometric authentication is required if either a custom biometric key
     * ({@link #getBiometryFactorRelatedKey()}) or biometric prompt data
     * ({@link #getBiometricPrompt()}) is present.
     *
     * @return {@code true} if biometric authentication should be used, {@code false} otherwise.
     */
    public boolean useBiometricFactor() {
        return biometricPrompt != null || useBiometry != null;
    }

    /**
     * @return Data for biometric prompt.
     */
    @Nullable
    public PowerAuthBiometricPrompt getBiometricPrompt() {
        return biometricPrompt;
    }
    /**
     * @return Biometry key data, or nil if biometry factor is not used.
     */
    @Nullable
    public SecureData getBiometryFactorRelatedKey() {
        return useBiometry;
    }

    /**
     * @return Password to be used for knowledge factor, or nil if knowledge factor is not used.
     */
    @Nullable
    public Password getPassword() {
        return password;
    }

    /**
     * @return If non-null, then custom key is specified for the possession factor.
     */
    @Nullable
    public SecureData getOverriddenPossessionKey() {
        return overriddenPossessionKey;
    }

    /**
     * Destroys any sensitive content, such as password stored in the authentication object.
     * After this call, the object becomes unusable for authentication operations.
     */
    public void destroy() {
        releaseSensitiveData(false);
    }

    // Internal interfaces

    /**
     * Calculate numeric value representing a combination of used factors.
     * @return Numeric value representing a combination of factors.
     */
    int getAuthenticationCodeFactorsMask() {
        int factors = 1;
        if (password != null) {
            factors |= 2;
        }
        if (useBiometry != null) {
            factors |= 4;
        }
        return factors;
    }

    /**
     * Validate usage of PowerAuthAuthentication object. If something doesn't match, then function
     * print warning to the debug console.
     * @param forPersist If true, then activation persist is expected.
     * @return false if object is created for the different purpose or is legacy constructed.
     */
    boolean validateAuthenticationUsage(boolean forPersist) {
        boolean result = validateAuthenticationUsageImpl(forPersist);
        if (!result && strictAuthenticationUsageValidation) {
            throw new IllegalArgumentException("Invalid PowerAuthAuthentication object provided");
        }
        return result;
    }

    /**
     * If set to true, then validateAuthenticationUsage() throws IllegalArgumentException().
     */
    private static boolean strictAuthenticationUsageValidation = false;

    /**
     * Enable or disable strict mode for {@link #validateAuthenticationUsage(boolean)} method.
     * If strict mode is enabled, then validation throws an error in case that validation fails.
     * This is useful only for PowerAuth SDK unit and integration testing.
     *
     * @param strict Enable or disable strict mode.
     */
    static void setStrictValidateAuthenticationUsage(boolean strict) {
        strictAuthenticationUsageValidation = strict;
    }

    /**
     * Validate usage of PowerAuthAuthentication object. If something doesn't match, then function
     * print warning to the debug console.
     * @param forPersist If true, then activation persist is expected.
     * @return false if object is created for the different purpose or is legacy constructed.
     */
    private boolean validateAuthenticationUsageImpl(boolean forPersist) {
        if (persistActivation == null) {
            PowerAuthLog.w("Using PowerAuthAuthentication object created with legacy constructor.");
            return false;
        } else {
            if (persistActivation != forPersist) {
                if (forPersist) {
                    PowerAuthLog.w("Using PowerAuthAuthentication object for a different purpose. The object to persist activation is expected.");
                } else {
                    PowerAuthLog.w("Using PowerAuthAuthentication object for a different purpose. The object for authentication code calculation is expected.");
                }
                return false;
            }
        }
        return true;
    }
}

File: proj-xcode/PowerAuth2ForExtensions/PowerAuthAuthentication.m
/**
 * Copyright 2021 Wultra s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// PA2_SHARED_SOURCE PowerAuth2ForWatch .
// PA2_SHARED_SOURCE PowerAuth2ForExtensions .

#import <PowerAuth2ForExtensions/PowerAuthAuthentication.h>
#import <PowerAuth2ForExtensions/PowerAuthKeychainAuthentication.h>
#import <PowerAuth2ForExtensions/PowerAuthLog.h>
#import "PowerAuthAuthentication+Private.h"

@implementation PowerAuthAuthentication
{
	NSInteger _objectUsage;
}

#define AUTH_FOR_COMMIT			1
#define AUTH_FOR_SIGN			2

- (id) initWithObjectUsage:(NSInteger)objectUsage
				  password:(NSString*)password
				  biometry:(BOOL)biometry
			biometryPrompt:(NSString*)biometryPrompt
		   biometryContext:(id)biometryContext
	   customPossessionKey:(NSData*)customPossessionKey
		 customBiometryKey:(NSData*)customBiometryKey
{
	self = [super init];
	if (self) {
		_objectUsage = objectUsage;
		_usePossession = YES;
		_usePassword = password;
		_useBiometry = biometry;
		_biometryPrompt = biometryPrompt;
		_biometryContext = biometryContext;
		_overridenPossessionKey = customPossessionKey;
		_overridenBiometryKey = customBiometryKey;
	}
	return self;
}

- (id) copyWithZone:(NSZone *)zone
{
	PowerAuthAuthentication * copy = [[[self class] allocWithZone:zone] init];
	if (copy) {
		copy->_objectUsage = _objectUsage;
		copy->_usePossession = _usePossession;
		copy->_useBiometry = _useBiometry;
		copy->_usePassword = _usePassword;
		copy->_biometryPrompt = _biometryPrompt;
		copy->_overridenPossessionKey = _overridenPossessionKey;
		copy->_overridenBiometryKey = _overridenBiometryKey;
#if PA2_HAS_LACONTEXT == 1
		copy->_biometryContext = _biometryContext;
#endif
	}
	return copy;
}

- (PowerAuthKeychainAuthentication *) keychainAuthentication
{
#if PA2_HAS_LACONTEXT == 1
	if (_biometryContext) {
		return [[PowerAuthKeychainAuthentication alloc] initWithContext:_biometryContext];
	}
#endif // PA2_HAS_LACONTEXT
	if (_biometryPrompt) {
		return [[PowerAuthKeychainAuthentication alloc] initWithPrompt:_biometryPrompt];
	}
	return nil;
}

#if DEBUG
- (NSString*) description
{
	NSString * usage_str;
	if (_objectUsage == AUTH_FOR_SIGN) {
		usage_str = @"for sign";
	} else if (_objectUsage == AUTH_FOR_COMMIT) {
		usage_str = @"for commit";
	} else {
		usage_str = @"legacy";
	}
	NSMutableArray * factors = [NSMutableArray arrayWithCapacity:3];
	if (_usePossession) {
		[factors addObject:@"possession"];
	}
	if (_usePassword) {
		[factors addObject:@"knowledge"];
	}
	if (_useBiometry) {
		[factors addObject:@"biometry"];
	}
	NSString * factors_str = [factors componentsJoinedByString:@"_"];
	NSMutableArray * info = [NSMutableArray array];
	if (_biometryPrompt) {
		[info addObject:@"+prompt"];
	}
#if PA2_HAS_LACONTEXT == 1
	if (_biometryContext) {
		[info addObject:@"+context"];
	}
#endif
	if (_overridenBiometryKey) {
		[info addObject:@"+extBK"];
	}
	if (_overridenPossessionKey) {
		[info addObject:@"+extPK"];
	}
	NSString * info_str = info.count == 0 ? @"" : [@", " stringByAppendingString:[info componentsJoinedByString:@" "]];
	return [NSString stringWithFormat:@"<PowerAuthAuthentication %@: %@%@>", usage_str, factors_str, info_str];
}
#endif

@end


@implementation PowerAuthAuthentication (EasyAccessors)

// MARK: - Commit, Possession + Knowledge

#if PA2_HAS_CORE_MODULE

+ (PowerAuthAuthentication*) commitWithPassword:(NSString*)password
{
	return [[PowerAuthAuthentication alloc] initWithObjectUsage:AUTH_FOR_COMMIT
													   password:password
													   biometry:NO
												 biometryPrompt:nil
												biometryContext:nil
											customPossessionKey:nil
											  customBiometryKey:nil];
}

+ (PowerAuthAuthentication*) commitWithPassword:(NSString*)password
							customPossessionKey:(NSData*)customPossessionKey
{
	return [[PowerAuthAuthentication alloc] initWithObjectUsage:AUTH_FOR_COMMIT
													   password:password
													   biometry:NO
												 biometryPrompt:nil
												biometryContext:nil
											customPossessionKey:customPossessionKey
											  customBiometryKey:nil];
}

// MARK: Commit, Possession + Knowledge + Biometry

+ (PowerAuthAuthentication*) commitWithPasswordAndBiometry:(NSString*)password
{
	return [[PowerAuthAuthentication alloc] initWithObjectUsage:AUTH_FOR_COMMIT
													   password:password
													   biometry:YES
												 biometryPrompt:nil
												biometryContext:nil
											customPossessionKey:nil
											  customBiometryKey:nil];
}

+ (PowerAuthAuthentication*) commitWithPasswordAndBiometry:(NSString*)password
										 customBiometryKey:(NSData*)customBiometryKey
									   customPossessionKey:(NSData*)customPossessionKey
{
	return [[PowerAuthAuthentication alloc] initWithObjectUsage:AUTH_FOR_COMMIT
													   password:password
													   biometry:YES
												 biometryPrompt:nil
												biometryContext:nil
											customPossessionKey:customPossessionKey
											  customBiometryKey:customBiometryKey];
}

#endif // PA2_HAS_CORE_MODULE


// MARK: - Signing, Possession only

+ (PowerAuthAuthentication *) possession
{
	return [[PowerAuthAuthentication alloc] initWithObjectUsage:AUTH_FOR_SIGN
													   password:nil
													   biometry:NO
												 biometryPrompt:nil
												biometryContext:nil
											customPossessionKey:nil
											  customBiometryKey:nil];
}

+ (PowerAuthAuthentication *) possessionWithCustomPossessionKey:(NSData*)customPossessionKey
{
	return [[PowerAuthAuthentication alloc] initWithObjectUsage:AUTH_FOR_SIGN
													   password:nil
													   biometry:NO
												 biometryPrompt:nil
												biometryContext:nil
											customPossessionKey:customPossessionKey
											  customBiometryKey:nil];
}

// MARK: Signing, Possession + Biometry

+ (PowerAuthAuthentication *) possessionWithBiometry
{
	return [[PowerAuthAuthentication alloc] initWithObjectUsage:AUTH_FOR_SIGN
													   password:nil
													   biometry:YES
												 biometryPrompt:nil
												biometryContext:nil
											customPossessionKey:nil
											  customBiometryKey:nil];
}

+ (PowerAuthAuthentication *) possessionWithBiometryPrompt:(NSString*)biometryPrompt
{
	return [[PowerAuthAuthentication alloc] initWithObjectUsage:AUTH_FOR_SIGN
													   password:nil
													   biometry:YES
												 biometryPrompt:biometryPrompt
												biometryContext:nil
											customPossessionKey:nil
											  customBiometryKey:nil];
}

+ (PowerAuthAuthentication *) possessionWithBiometryPrompt:(NSString*)biometryPrompt
									   customPossessionKey:(NSData*)customPossessionKey
{
	return [[PowerAuthAuthentication alloc] initWithObjectUsage:AUTH_FOR_SIGN
													   password:nil
													   biometry:YES
												 biometryPrompt:biometryPrompt
												biometryContext:nil
											customPossessionKey:customPossessionKey
											  customBiometryKey:nil];
}

+ (PowerAuthAuthentication *) possessionWithBiometryWithCustomBiometryKey:(NSData*)customBiometryKey
													  customPossessionKey:(NSData*)customPossessionKey
{
	return [[PowerAuthAuthentication alloc] initWithObjectUsage:AUTH_FOR_SIGN
													   password:nil
													   biometry:YES
												 biometryPrompt:nil
												biometryContext:nil
											customPossessionKey:customPossessionKey
											  customBiometryKey:customBiometryKey];
}

#if PA2_HAS_LACONTEXT == 1
+ (PowerAuthAuthentication *) possessionWithBiometryContext:(LAContext *)context
{
	return [[PowerAuthAuthentication alloc] initWithObjectUsage:AUTH_FOR_SIGN
													   password:nil
													   biometry:NO
												 biometryPrompt:nil
												biometryContext:context
											customPossessionKey:nil
											  customBiometryKey:nil];
}
+ (PowerAuthAuthentication *) possessionWithBiometryContext:(LAContext*)context
										customPossessionKey:(NSData*)customPossessionKey
{
	return [[PowerAuthAuthentication alloc] initWithObjectUsage:AUTH_FOR_SIGN
													   password:nil
													   biometry:NO
												 biometryPrompt:nil
												biometryContext:context
											customPossessionKey:customPossessionKey
											  customBiometryKey:nil];
}
#endif // PA2_HAS_LACONTEXT

// MARK: Signing, Possession + Knowledge

+ (PowerAuthAuthentication *) possessionWithPassword:(NSString *)password
{
	return [[PowerAuthAuthentication alloc] initWithObjectUsage:AUTH_FOR_SIGN
													   password:password
													   biometry:NO
												 biometryPrompt:nil
												biometryContext:nil
											customPossessionKey:nil
											  customBiometryKey:nil];
}

+ (PowerAuthAuthentication *) possessionWithKnowledge:(NSString*)password
								  customPossessionKey:(NSData*)customPossessionKey
{
	return [[PowerAuthAuthentication alloc] initWithObjectUsage:AUTH_FOR_SIGN
													   password:password
													   biometry:NO
												 biometryPrompt:nil
												biometryContext:nil
											customPossessionKey:customPossessionKey
											  customBiometryKey:nil];
}

#pragma mark - Deprecated

// PA2_DEPRECATED(1.7.0)
+ (PowerAuthAuthentication *) possessionWithBiometryWithPrompt:(NSString *)biometryPrompt
{
	return [self possessionWithBiometryPrompt:biometryPrompt];
}
// PA2_DEPRECATED(1.7.0)
+ (PowerAuthAuthentication *) possessionWithPasswordDeprecated:(NSString*)password
{
	return [self possessionWithPassword:password];
}
@end


@implementation PowerAuthAuthentication (Private)

- (NSInteger) signatureFactorMask
{
	NSUInteger result = 0;
	if (_usePossession) result |= 1;
	if (_usePassword)   result |= 2;
	if (_useBiometry)   result |= 4;
	return result;
}

- (BOOL) validateUsage:(BOOL)forCommit
{
	if (_objectUsage == 0) {
		PowerAuthLog(@"WARNING: Using PowerAuthAuthentication object created with legacy constructor.");
		return NO;
	}
	if (forCommit != (_objectUsage == AUTH_FOR_COMMIT)) {
		if (forCommit) {
			PowerAuthLog(@"WARNING: Using PowerAuthAuthentication object for a different purpose. The object for activation commit is expected.");
		} else {
			PowerAuthLog(@"WARNING: Using PowerAuthAuthentication object for a different purpose. The object for signature calculation is expected.");
		}
		return NO;
	}
	return YES;
}

@end
